modules/ml/src/main/java/org/apache/ignite/ml/composition/stacking/StackedDatasetTrainer.java

```diff
@@ -0,0 +1,410 @@
+        IgniteFunction<IS, IA> submodelInput2AggregatingInputConverter) {
+        this.submodelInput2AggregatingInputConverter = submodelInput2AggregatingInputConverter;
+
+        return this;
+    }
+
+    /**
+     * Drop original features during training and inference.
+     *
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withOriginalFeaturesDropped() {
+        this.submodelInput2AggregatingInputConverter = null;
+
+        return this;
+    }
+
+    /**
+     * Set function used for conversion of submodel output to {@link Vector}. This function is used during
+     * building of dataset for training aggregator model. This dataset is augmented with results of submodels
+     * converted to {@link Vector}.
+     *
+     * @param submodelOutput2VectorConverter Function used for conversion of submodel output to {@link Vector}.
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withSubmodelOutput2VectorConverter(
+        IgniteFunction<IA, Vector> submodelOutput2VectorConverter) {
+        this.submodelOutput2VectorConverter = submodelOutput2VectorConverter;
+
+        return this;
+    }
+
+    /**
+     * Set function used for conversion of {@link Vector} to submodel input. This function is used during
+     * building of dataset for training aggregator model. This dataset is augmented with results of submodels
+     * applied to {@link Vector}s in original dataset.
+     *
+     * @param vector2SubmodelInputConverter Function used for conversion of {@link Vector} to submodel input.
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withVector2SubmodelInputConverter(
+        IgniteFunction<Vector, IS> vector2SubmodelInputConverter) {
+        this.vector2SubmodelInputConverter = vector2SubmodelInputConverter;
+
+        return this;
+    }
+
+    /**
+     * Specify aggregator trainer.
+     *
+     * @param aggregatorTrainer Aggregator trainer.
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withAggregatorTrainer(DatasetTrainer<AM, L> aggregatorTrainer) {
+        this.aggregatorTrainer = aggregatorTrainer;
+
+        return this;
+    }
+
+    /**
+     * Specify binary operator used to merge submodels outputs to one.
+     *
+     * @param merger Binary operator used to merge submodels outputs to one.
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withAggregatorInputMerger(IgniteBinaryOperator<IA> merger) {
+        this.aggregatingInputMerger = merger;
+
+        return this;
+    }
+
+    /**
+     * Adds submodel trainer along with converters needed on training and inference stages.
+     *
+     * @param trainer Submodel trainer.
+     * @return This object.
+     */
+    public <M1 extends Model<IS, IA>> StackedDatasetTrainer<IS, IA, O, AM, L> addTrainer(
```

Comment:
Reason is to convert everything into `DatasetTrainer<Model<IS, IA>>` (in contrast to `DatasetTrainer<? extends Model<IS, IA>>`) to make work with list of `submodelTrainers` less painful. This is unsafe conversion, but since we have control of all `sumbodelTrainers` list usages inside our class, it's IMHO a reasanoble tradeoff.